### DISCUSSION TITLE
How to set and use `readyPromise` ?

### DISCUSSION THREAD

**Discussion by bric3** [excalidraw/excalidraw#3755] [Q&A] [answered]
Hi everyone,

I'm pasting the verbatim content of this the question I asked in https://github.com/excalidraw/excalidraw/issues/3752 as advised by @ad1992.

----

Disclaimer: I am not a JS, React developer, meaning I ignore many things of this world.

I have seen this sandbox : https://codesandbox.io/s/excalidraw-ehlz3?file=/src/App.js

I have seen the documentation of the `ref` property here : 

* https://github.com/excalidraw/excalidraw/tree/master/src/packages/excalidraw#props
* https://github.com/excalidraw/excalidraw/tree/master/src/packages/excalidraw#ref

Which links to this code snippet : https://github.com/excalidraw/excalidraw/blob/master/src/utils.ts#L343-L357

And there's also the readyPromise property which I don't quite get
https://github.com/excalidraw/excalidraw/tree/master/src/packages/excalidraw#readypromise

But when declaring the promise this way didn't work :

```jsx
import React from "react";
import Excalidraw, {
  ...
} from "@excalidraw/excalidraw";

export default function App() {
    const excalidrawRef = React.useRef(null);

    ...

    return (
        <div className="excalidraw-wrapper">
            <Excalidraw
                ref={excalidrawRef}
                onChange={(elements, state) =>
                    console.log("Elements :", elements, "State : ", state)
                }
                onCollabButtonClick={() =>
                    window.alert("Not supported")
                }
                readyPromise={() => {
                    console.log("excalidraw ready")
                }}
                initialData={{ appState: {
                        exportEmbedScene: true
                    }
                }}
                UIOptions={{canvasActions: {loadScene: false}}}
            />
        </div>
    );
}
```

I expected the lambda function to be called when Excalidraw is loaded and ready. But it didn't happen. There's something that's eluding me.

Thank you in advance for shedding the light on this!

**Comment by ad1992** [comment, marked as answer]
@bric3 thanks for asking the question!
`readyPromise` is not a prop so you cannot pass it as a prop to the component.
Before diving into `readyPromise` let me explain how you can pass `refs`.

You can pass `ref` to the excalidraw component in one of the following ways 👇🏻 
1. [`createRef` ](https://reactjs.org/docs/refs-and-the-dom.html#creating-refs) / [`useRef`](https://reactjs.org/docs/hooks-reference.html#useref), the difference between the two is `createRef` will return a new reference every time the component re-renders so you will want to cache the reference when using the `class` component, and the same can be achieved using `useRef` in the functional component.
2. using `callbackRef`(https://reactjs.org/docs/refs-and-the-dom.html#callback-refs), will make sure the callback is triggered only after the component mounts which I think should suffice your use case, added an [example with callback ref](https://codesandbox.io/s/excalidraw-callback-ref-stmhc?file=/src/App.js) for the same.
3. passing plain object 
<pre>
const excalidrawRef = { current: { readyPromise: <a href="https://github.com/excalidraw/excalidraw/blob/master/src/utils.ts#L343-L357">resolvablePromise</a>}}
</pre>
Since an object is passed here so the `readyPromise` is resolved as soon as the component is mounted. Most of the time you will not need this unless you have a specific use case where you can't pass the `ref `in the react way and want to do some action on the host when this promise resolves, added an [example with resolvable promise](https://codesandbox.io/s/eexcalidraw-resolvable-promise-d0qg3?file=/src/App.js)

Additionally `ready` in the `ref` API is set to true once the component is loaded.

**Reply by bric3** [reply]
Hi @ad1992 Great explanation!

And the code sample are really useful !Thanks a lot. They really should be part of the documentation.

Indeed I think `useCallback` is likely what I need. I'll try that tomorrow.

**Reply by ad1992** [reply]
> And the code sample are really useful !Thanks a lot. They really should be part of the documentation.

Sure, will add in the docs as well.

**Reply by bric3** [reply]
I noticed I needed to use the component reference as well as notify the host about it's readiness.
Does this use of `useCallback` looks reasonable ? 

```javascript
export default function App() {
    const excalidrawApiRef = React.useRef(null);
    const excalidrawRef = React.useCallback((excalidrawApi) => {
        excalidrawApiRef.current = excalidrawApi;
        dispatchToPlugin({ type: "ready" })
    }, []);

   ... later use excalidrawApiRef.current

    return return (
        <div className="excalidraw-wrapper">
            <Excalidraw
                ref={excalidrawRef}
                ...
    );
};
```

**Reply by ad1992** [reply]
sure you can cache the API using `useRef` so it persists for the full lifetime of the component but `excalidrawApiRef.current` will be `null` until the callback is triggered so you will need to add a safe check to check `excalidrawApiRef.current` exists.

**Reply by bric3** [reply]
Yes that's expected, the host should only call apis once the component is ready !

**Reply by bric3** [reply]
Thank you so much for your time !

**Reply by ad1992** [reply]
glad to help :)
